Fix instance eval object

MattiasFredriksson opened the pull request:
Description
----
Fix for #2446. Geometry node (GN) instances are currently exporting the 'evaluated depsgraph object' rather then the instantiated (or 'original') mesh data block. Proposed changes will instead export the original un-evaluated mesh when the object has no modifiers applied.

Initially did not consider modified instantiated objects, generally the evaluated object will be exported. While a slight improvement it is a bit tricky as it would be an edge case. Due to how 'apply modifier' option works it will still export separate mesh entries for those cases even if they are both evaluations of the same object.

Changes
----
- Export the original GN instantiated mesh object when possible (it has no modifiers)
- Unittest with .blend validating that there are no additional `meshes` entries exported for the unmodified case and copies are exported in the case with modifiers!
- Tested on Blender 4.4 beta version.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/KhronosGroup/glTF-Blender-IO?pullRequest=2490) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/KhronosGroup/glTF-Blender-IO?pullRequest=2490) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/KhronosGroup/glTF-Blender-IO?pullRequest=2490) it.</sub>

julienduroure commented:
Hello,
Thanks for the PR.
I will have a look on it soon

MattiasFredriksson commented:
No worries, it is an improvement to current behavior but I think it's more a bandaid and doesn't directly solve the underlying problem.

The core issue is mainly that the importer has two options that are handled independently: 'Apply Modifier' and 'GN Instances'. Selecting one and not the other exports parts of the evaluation but not the other part (e.g. the GN modified mesh output + GN instances). What happends are these issue where one part is 'evaluated' and the other part isn't.

Addressing those issues are a bit complicated though because the mesh node is re-evaluated at export and is another copy. Improving all of those behaviors is a bit of a larger issue so it would be a good idea to get your opinion on the topic and these changes before commiting to trying to do a larger change.

julienduroure commented:
We are currently working with Blender core dev to get an exporter evaluation graph that get all data needed for geometry export, making this PR and current GN related code no more relevant.
Let's keep this PR open until I get more info about it, and until I tested it

MattiasFredriksson commented:
Yes [this change](https://projects.blender.org/blender/blender/pulls/135318) should make things better. Will close this as I intended to do so but forgot about it!
